explorer/interpreter/interpreter.cpp

```diff
@@ -1141,68 +1188,82 @@ auto Interpreter::StepExp() -> ErrorOr<Success> {
+            std::optional<const Value*> type_result;
+            if (!isa<InterfaceType, NamedConstraintType, ConstraintType>(
+                    act.results()[0])) {
+              type_result = act.results()[0];
+            }
+            MemberName* member_name = arena_->New<MemberName>(
+                type_result, found_in_interface, member_name_type->member());
+            return todo_.FinishAction(member_name);
           }
-          MemberName* member_name = arena_->New<MemberName>(
-              type_result, found_in_interface, member_name_type->member());
-          return todo_.FinishAction(member_name);
         } else {
           // The result is the value of the named field, such as in
           // `value.field_name`. Extract the value within the given object.
-          std::optional<Nonnull<const Witness*>> witness;
-          if (access.impl().has_value()) {
-            witness = cast<Witness>(act.results()[1]);
-          }
-          ElementPath::Component member(&access.member(), found_in_interface,
-                                        witness);
-          const Value* aggregate;
-          if (access.is_type_access()) {
-            CARBON_ASSIGN_OR_RETURN(
-                aggregate, InstantiateType(&access.object().static_type(),
-                                           access.source_loc()));
-          } else if (const auto* lvalue = dyn_cast<LValue>(act.results()[0])) {
-            CARBON_ASSIGN_OR_RETURN(
-                aggregate,
-                this->heap_.Read(lvalue->address(), exp.source_loc()));
+          if (act.pos() == 1) {
+            // Next, if we're accessing an interface member, evaluate the `impl`
+            // expression to find the corresponding witness.
+            if (auto impl_has_value = access.impl().has_value()) {
+              return todo_.Spawn(
+                  std::make_unique<WitnessAction>(access.impl().value()));
+            } else {
+              return todo_.RunAgain();
+            }
+          } else if (auto found_in_interface = access.found_in_interface();
+                     found_in_interface && act.pos() == 2) {
+            return todo_.Spawn(std::make_unique<TypeInstantiationAction>(
+                *found_in_interface, exp.source_loc()));
           } else {
-            aggregate = act.results()[0];
+            if (found_in_interface) {
+              found_in_interface = cast<InterfaceType>(act.results().back());
+            }
+            std::optional<Nonnull<const Witness*>> witness;
+            if (access.impl().has_value()) {
+              witness = cast<Witness>(act.results()[1]);
+            }
+            ElementPath::Component member(&access.member(), found_in_interface,
+                                          witness);
+            const Value* aggregate;
+            if (access.is_type_access()) {
+              CARBON_ASSIGN_OR_RETURN(
+                  aggregate, InstantiateType(&access.object().static_type(),
```

Comment:
Ha, I suppose that's another opportunity to use the Action you created. Hopefully that makes the rest a bit more readable / symmetric, GH makes it hard to parse the diff.

Reply:
Sorry, I missed it by mistake, making it similar to how I did it for case `CompoundMemberAccessExpression`